File: python/metatomic_torch/tests/units.py
import math

import ase.units
import pytest
import torch

from metatomic.torch import ModelOutput, unit_conversion_factor


# 3-arg C++ op (deprecated, but still registered for backward compat)
_unit_conversion_factor_3arg = torch.ops.metatomic.unit_conversion_factor


# ---- Backward compat: 3-arg C++ op still works (with deprecation warning) ----


def test_conversion_length_3arg(capfd):
    length_angstrom = 1.0
    length_nm = (
        _unit_conversion_factor_3arg("length", "angstrom", "nm") * length_angstrom
    )
    assert length_nm == pytest.approx(0.1)

    captured = capfd.readouterr()
    assert captured.out == ""
    message = (
        "the 3-argument unit_conversion_factor(quantity, from, to) is "
        "deprecated; use the 2-argument unit_conversion_factor(from, to) "
        "instead. The quantity parameter is no longer needed."
    )
    assert message in captured.err


def test_conversion_energy_3arg():
    energy_ev = 1.0
    energy_mev = _unit_conversion_factor_3arg("energy", "ev", "mev") * energy_ev
    assert energy_mev == pytest.approx(1000.0)


# ---- 2-arg API ----


def test_conversion_length():
    assert unit_conversion_factor("angstrom", "nm") == pytest.approx(0.1)
    assert unit_conversion_factor("angstrom", "Bohr") == pytest.approx(
        1.8897259886, rel=1e-6
    )
    assert unit_conversion_factor("Angstrom", "meter") == pytest.approx(
        1e-10, rel=1e-10
    )


def test_conversion_energy():
    assert unit_conversion_factor("eV", "meV") == pytest.approx(1000.0)
    assert unit_conversion_factor("eV", "Hartree") == pytest.approx(0.0367493, rel=1e-3)


# ---- 2-arg API with keyword arguments ----


def test_conversion_length_kwargs():
    # Test keyword argument form: from_unit and to_unit
    assert unit_conversion_factor(from_unit="angstrom", to_unit="nm") == pytest.approx(
        0.1
    )
    assert unit_conversion_factor(
        from_unit="angstrom", to_unit="Bohr"
    ) == pytest.approx(1.8897259886, rel=1e-6)
    assert unit_conversion_factor(
        from_unit="Angstrom", to_unit="meter"
    ) == pytest.approx(1e-10, rel=1e-10)


def test_conversion_energy_kwargs():
    # Test keyword argument form: from_unit and to_unit
    assert unit_conversion_factor(from_unit="eV", to_unit="meV") == pytest.approx(
        1000.0
    )
    assert unit_conversion_factor(from_unit="eV", to_unit="Hartree") == pytest.approx(
        0.0367493, rel=1e-3
    )


# ---- Mixed positional and keyword arguments ----


def test_conversion_mixed_args():
    # Test mixed positional and keyword arguments
    # First arg positional, second as keyword
    assert unit_conversion_factor("angstrom", to_unit="nm") == pytest.approx(0.1)
    # First arg as keyword, second positional
    assert unit_conversion_factor(from_unit="angstrom", to_unit="nm") == pytest.approx(
        0.1
    )


def test_units_vs_ase():
    assert unit_conversion_factor("angstrom", "bohr") == pytest.approx(
        ase.units.Ang / ase.units.Bohr, rel=1e-6
    )
    assert unit_conversion_factor("angstrom", "nm") == pytest.approx(
        ase.units.Ang / ase.units.nm
    )
    assert unit_conversion_factor("angstrom", "nanometer") == pytest.approx(
        ase.units.Ang / ase.units.nm
    )

    assert unit_conversion_factor("ev", "Hartree") == pytest.approx(
        ase.units.eV / ase.units.Hartree, rel=1e-6
    )
    kcal_mol = ase.units.kcal / ase.units.mol
    assert unit_conversion_factor("ev", "kcal/mol") == pytest.approx(
        ase.units.eV / kcal_mol, rel=1e-4
    )
    kJ_mol = ase.units.kJ / ase.units.mol
    assert unit_conversion_factor("ev", "kJ/mol") == pytest.approx(
        ase.units.eV / kJ_mol, rel=1e-4
    )


# ---- Compound expressions ----


def test_compound_expressions():
    # Force: eV/Angstrom -> Hartree/Bohr
    conv = unit_conversion_factor("eV/Angstrom", "Hartree/Bohr")
    assert conv == pytest.approx(0.0194469, rel=1e-3)

    # kJ/mol -> kcal/mol
    conv = unit_conversion_factor("kJ/mol", "kcal/mol")
    assert conv == pytest.approx(1000.0 / 4184.0, rel=1e-6)

    # Pressure identity
    conv = unit_conversion_factor("eV/Angstrom^3", "eV/A^3")
    assert conv == pytest.approx(1.0)


# ---- Fractional powers ----


def test_fractional_powers():
    # (eV*u)^(1/2) -> u*A/fs
    conv = unit_conversion_factor("(eV*u)^(1/2)", "u*A/fs")
    ev_si = 1.602176634e-19
    u_si = 1.66053906660e-27
    a_si = 1e-10
    fs_si = 1e-15
    expected = math.sqrt(ev_si * u_si) / (u_si * a_si / fs_si)
    assert conv == pytest.approx(expected, rel=1e-3)

    # (eV/u)^(1/2) -> A/fs
    conv = unit_conversion_factor("(eV/u)^(1/2)", "A/fs")
    expected = math.sqrt(ev_si / u_si) / (a_si / fs_si)
    assert conv == pytest.approx(expected, rel=1e-3)


# ---- Dimension mismatch ----


def test_dimension_mismatch():
    with pytest.raises((ValueError, RuntimeError), match="dimension mismatch"):
        unit_conversion_factor("eV", "Angstrom")


# ---- Unknown unit ----


def test_unknown_unit():
    with pytest.raises((ValueError, RuntimeError), match="unknown unit"):
        unit_conversion_factor("foobar", "eV")


# ---- Empty string ----


def test_empty_string():
    assert unit_conversion_factor("", "eV") == 1.0
    assert unit_conversion_factor("eV", "") == 1.0
    assert unit_conversion_factor("", "") == 1.0


# ---- Overflow/underflow handling ----


def test_overflow_exponentiation():
    # Test overflow with extreme exponent on unit with small SI factor
    # u (atomic mass unit) has factor 1.66e-27, so u^(-50) overflows
    with pytest.raises((ValueError, RuntimeError), match="overflows"):
        unit_conversion_factor("u^(-50)", "u^(-50)")

    # eV has factor 1.6e-19, so eV^(-100) overflows
    with pytest.raises((ValueError, RuntimeError), match="overflows"):
        unit_conversion_factor("eV^(-100)", "eV^(-100)")


def test_overflow_multiplication():
    # Test overflow with multiplication of units with extreme factors
    # u^(-25) * u^(-25) = u^(-50), which overflows
    with pytest.raises((ValueError, RuntimeError), match="overflows"):
        unit_conversion_factor("u^(-25) * u^(-25)", "u^(-50)")


def test_overflow_division():
    # Test overflow with division creating extreme factor (same dimension)
    # u^(-25) / u^25 = u^(-50), which overflows (u has factor 1.66e-27)
    with pytest.raises((ValueError, RuntimeError), match="overflows"):
        unit_conversion_factor("u^(-25)", "u^25")


# ---- Valid units (ModelOutput creation still works) ----


def test_valid_units():
    # just checking that all of these are valid
    ModelOutput(quantity="length", unit="A")
    ModelOutput(quantity="length", unit="Angstrom")
    ModelOutput(quantity="length", unit="Bohr")
    ModelOutput(quantity="length", unit="meter")
    ModelOutput(quantity="length", unit=" centimeter")
    ModelOutput(quantity="length", unit="cm")
    ModelOutput(quantity="length", unit="millimeter")
    ModelOutput(quantity="length", unit="mm")
    ModelOutput(quantity="length", unit=" micrometer")
    ModelOutput(quantity="length", unit="um")
    ModelOutput(quantity="length", unit="µm")
    ModelOutput(quantity="length", unit="nanometer")
    ModelOutput(quantity="length", unit="nm ")

    ModelOutput(quantity="energy", unit="eV")
    ModelOutput(quantity="energy", unit="meV")
    ModelOutput(quantity="energy", unit="Hartree")
    ModelOutput(quantity="energy", unit="kcal /  mol ")
    ModelOutput(quantity="energy", unit="kJ/mol")
    ModelOutput(quantity="energy", unit="Joule")
    ModelOutput(quantity="energy", unit="J")
    ModelOutput(quantity="energy", unit="Rydberg")
    ModelOutput(quantity="energy", unit="Ry")

    ModelOutput(quantity="force", unit="eV/Angstrom")
    ModelOutput(quantity="force", unit="eV/A")

    ModelOutput(quantity="pressure", unit="eV/Angstrom^3")
    ModelOutput(quantity="pressure", unit="eV/A^3")

    ModelOutput(quantity="momentum", unit="u * A/ fs")
    ModelOutput(quantity="momentum", unit=" (eV*u )^(1/ 2 )")

    ModelOutput(quantity="velocity", unit="A/fs")
    ModelOutput(quantity="velocity", unit="A/s")


# ---- Accumulated floating-point error with fractional powers ----


def test_fractional_power_accumulated_error():
    # Test that (eV*u)^(1/3) then ^3 equals eV*u within tolerance
    # This verifies the 1e-10 dimension comparison tolerance handles
    # floating-point accumulation from fractional exponents
    conv = unit_conversion_factor("((eV*u)^(1/3))^3", "eV*u")
    assert conv == pytest.approx(1.0, rel=1e-6)

    # Test with 1/2 then ^2
    conv2 = unit_conversion_factor("((eV*u)^(1/2))^2", "eV*u")
    assert conv2 == pytest.approx(1.0, rel=1e-6)

    # Test nested: ((eV^(1/2))^2) should equal eV
    conv3 = unit_conversion_factor("(eV^(1/2))^2", "eV")
    assert conv3 == pytest.approx(1.0, rel=1e-6)


# ---- Time units ----


def test_time_units():
    assert unit_conversion_factor("s", "fs") == pytest.approx(1e15)
    assert unit_conversion_factor("second", "ps") == pytest.approx(1e12)
    assert unit_conversion_factor("ns", "fs") == pytest.approx(1e6)
    assert unit_conversion_factor("us", "ns") == pytest.approx(1e3)
    assert unit_conversion_factor("ms", "us") == pytest.approx(1e3)


# ---- Mass and charge units ----


def test_mass_units():
    assert unit_conversion_factor("u", "kg") == pytest.approx(1.66053906660e-27)
    assert unit_conversion_factor("dalton", "u") == pytest.approx(1.0)
    assert unit_conversion_factor("electron_mass", "kg") == pytest.approx(
        9.1093837015e-31, rel=1e-6
    )
    assert unit_conversion_factor("m_e", "electron_mass") == pytest.approx(1.0)


def test_charge_units():
    assert unit_conversion_factor("e", "Coulomb") == pytest.approx(1.602176634e-19)
    assert unit_conversion_factor("e", "C") == pytest.approx(1.602176634e-19)


def test_derived_constants():
    # hbar has dimensions of energy * time = M L^2 T^-1
    # hbar = 1.054571817e-34 J*s
    conv = unit_conversion_factor("hbar", "Joule * second")
    assert conv == pytest.approx(1.054571817e-34, rel=1e-6)


# ---- Micro sign for microsecond ----


def test_micro_sign_microsecond():
    # µs -> ns (microsecond via micro sign)
    assert unit_conversion_factor("µs", "ns") == pytest.approx(
        unit_conversion_factor("us", "ns")
    )


# ---- Quantity-unit mismatch ----


def test_quantity_unit_mismatch():
    # energy quantity with force unit
    with pytest.raises((ValueError, RuntimeError), match="incompatible with quantity"):
        ModelOutput(quantity="energy", unit="eV/A")

    # force quantity with energy unit
    with pytest.raises((ValueError, RuntimeError), match="incompatible with quantity"):
        ModelOutput(quantity="force", unit="eV")

    # length quantity with pressure unit
    with pytest.raises((ValueError, RuntimeError), match="incompatible with quantity"):
        ModelOutput(quantity="length", unit="eV/A^3")


# ---- Deprecation warning for 3-arg C++ op ----


def test_3arg_deprecation_warning():
    # The 3-arg C++ op emits a torch warning (not a Python DeprecationWarning)
    # on first call. We just verify the call succeeds and returns the right value.
    result = _unit_conversion_factor_3arg("energy", "eV", "meV")
    assert result == pytest.approx(1000.0)


def test_3arg_kwargs_backward_compat():
    # Verify that the old kwargs-based calling convention still works
    result = _unit_conversion_factor_3arg(
        quantity="energy", from_unit="eV", to_unit="meV"
    )
    assert result == pytest.approx(1000.0)


def test_3arg_all_calling_conventions():
    """Test all supported calling conventions for the deprecated 3-arg form."""
    # 3 positional arguments
    assert _unit_conversion_factor_3arg("energy", "eV", "meV") == pytest.approx(1000.0)
    assert _unit_conversion_factor_3arg("length", "angstrom", "nm") == pytest.approx(
        0.1
    )

    # 2 positional + to_unit keyword
    assert _unit_conversion_factor_3arg("energy", "eV", to_unit="meV") == pytest.approx(
        1000.0
    )
    assert _unit_conversion_factor_3arg(
        "length", "angstrom", to_unit="nm"
    ) == pytest.approx(0.1)

    # 1 positional + from_unit + to_unit keywords
    assert _unit_conversion_factor_3arg(
        "energy", from_unit="eV", to_unit="meV"
    ) == pytest.approx(1000.0)
    assert _unit_conversion_factor_3arg(
        "length", from_unit="angstrom", to_unit="nm"
    ) == pytest.approx(0.1)

    # All keywords
    assert _unit_conversion_factor_3arg(
        quantity="energy", from_unit="eV", to_unit="meV"
    ) == pytest.approx(1000.0)
    assert _unit_conversion_factor_3arg(
        quantity="length", from_unit="angstrom", to_unit="nm"
    ) == pytest.approx(0.1)


def test_2arg_all_calling_conventions():
    """Test all supported calling conventions for the 2-arg form."""
    # 2 positional arguments
    assert unit_conversion_factor("eV", "meV") == pytest.approx(1000.0)
    assert unit_conversion_factor("angstrom", "nm") == pytest.approx(0.1)

    # All keywords
    assert unit_conversion_factor(from_unit="eV", to_unit="meV") == pytest.approx(
        1000.0
    )
    assert unit_conversion_factor(from_unit="angstrom", to_unit="nm") == pytest.approx(
        0.1
    )

    # Mixed: first positional, second keyword
    assert unit_conversion_factor("eV", to_unit="meV") == pytest.approx(1000.0)
    assert unit_conversion_factor("angstrom", to_unit="nm") == pytest.approx(0.1)


def test_3arg_error_cases():
    """Test error handling for the deprecated 3-arg form."""
    # 2 positional args without to_unit keyword is treated as 2-arg form
    # This will fail with "unknown unit" since "energy" is not a valid unit expression
    with pytest.raises((ValueError, RuntimeError), match="unknown unit"):
        _unit_conversion_factor_3arg(
            "energy", "eV"
        )  # treated as 2-arg: from="energy", to="eV"

    # 1 positional arg is treated as 2-arg form with missing to_unit
    with pytest.raises(
        RuntimeError, match="unit_conversion_factor requires 2 arguments"
    ):
        _unit_conversion_factor_3arg(
            "energy"
        )  # treated as 2-arg: from="energy", to=None

    # quantity keyword alone without from_unit/to_unit is treated as 2-arg form
    # _0=None, from_unit="energy", to_unit=None → fails 2-arg validation
    with pytest.raises(
        RuntimeError, match="unit_conversion_factor requires 2 arguments"
    ):
        _unit_conversion_factor_3arg(quantity="energy")


def test_2arg_error_cases():
    """Test error handling for the 2-arg form."""
    # Missing arguments in 2-arg form
    with pytest.raises(
        RuntimeError,
        match="unit_conversion_factor requires 2 arguments: from_unit and to_unit",
    ):
        unit_conversion_factor("eV")  # missing to_unit

    with pytest.raises(
        RuntimeError,
        match="unit_conversion_factor requires 2 arguments: from_unit and to_unit",
    ):
        unit_conversion_factor()  # missing both

    # Partial kwargs in 2-arg form
    with pytest.raises(
        RuntimeError,
        match="unit_conversion_factor requires 2 arguments: from_unit and to_unit",
    ):
        unit_conversion_factor(from_unit="eV")  # missing to_unit


# ---- TorchScript compatibility ----


def test_torchscript_unit_conversion():
    """unit_conversion_factor must be callable from TorchScript."""

    @torch.jit.script
    def convert(from_unit: str, to_unit: str) -> float:
        return torch.ops.metatomic.unit_conversion_factor(from_unit, to_unit)

    assert convert("eV", "meV") == pytest.approx(1000.0)
    assert convert("angstrom", "nm") == pytest.approx(0.1)


# ---- Multithreading tests ----


def test_thread_local_cache():
    """Test that thread-local cache works correctly across multiple threads."""
    import threading

    num_threads = 10
    iterations_per_thread = 100
    results = [None] * (num_threads * iterations_per_thread)

    def worker(thread_id):
        for i in range(iterations_per_thread):
            factor = unit_conversion_factor("eV", "meV")
            results[thread_id * iterations_per_thread + i] = factor

    threads = []
    for t in range(num_threads):
        thread = threading.Thread(target=worker, args=(t,))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    # All results should be consistent (1000.0 for eV -> meV)
    for result in results:
        assert result == pytest.approx(1000.0, rel=1e-10)


def test_concurrent_different_conversions():
    """Test that different threads can safely convert different units."""
    import threading

    num_threads = 5
    results = [None] * num_threads

    conversions = [
        ("eV", "meV"),  # energy
        ("angstrom", "bohr"),  # length
        ("fs", "ps"),  # time
        ("u", "kg"),  # mass
        ("eV/A", "Hartree/Bohr"),  # force
    ]

    expected = [
        1000.0,  # eV -> meV
        1.8897259886,  # angstrom -> bohr
        0.001,  # fs -> ps
        1.66053906660e-27,  # u -> kg
        0.0194469,  # eV/A -> Hartree/Bohr
    ]

    def worker(thread_id):
        for _ in range(50):
            results[thread_id] = unit_conversion_factor(
                conversions[thread_id][0], conversions[thread_id][1]
            )

    threads = []
    for t in range(num_threads):
        thread = threading.Thread(target=worker, args=(t,))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    # Check each thread got the correct result
    for t in range(num_threads):
        assert results[t] == pytest.approx(expected[t], rel=1e-4)

File: metatomic-torch/tests/units.cpp
#include <cmath>
#include <string>

#include <torch/torch.h>

#include "metatomic/torch.hpp"

#include <catch.hpp>
using Catch::Detail::Approx;
using Catch::Matchers::Contains;
using Catch::Matchers::StartsWith;

class DeprecationWarningHandler : public torch::WarningHandler {
public:
    std::vector<std::string> messages;

    void process(const torch::Warning& warning) override {
        messages.push_back(warning.msg());
    }
};

// check that two doubles are approximately equal within a certain number of ULPs
static bool check_equal_ulp(double a, double b, size_t ulp = 4) {
    if (std::isnan(a) || std::isnan(b)) {
        return std::isnan(a) && std::isnan(b);
    } else if (std::isinf(a) || std::isinf(b)) {
        return std::isinf(a) && std::isinf(b);
    } else {
        double next_a = a;
        double prev_a = a;
        for (size_t i = 0; i < ulp; ++i) {
            next_a = std::nextafter(next_a, std::numeric_limits<double>::infinity());
            prev_a = std::nextafter(prev_a, -std::numeric_limits<double>::infinity());
        }

        double next_b = b;
        double prev_b = b;
        for (size_t i = 0; i < ulp; ++i) {
            next_b = std::nextafter(next_b, std::numeric_limits<double>::infinity());
            prev_b = std::nextafter(prev_b, -std::numeric_limits<double>::infinity());
        }

        return (a >= prev_b && a <= next_b) && (b >= prev_a && b <= next_a);
    }
}

#define CHECK_APPROX_ULP(a, b) do { if (!check_equal_ulp(a, b)) { CHECK(a == b); } } while(0)


TEST_CASE("Simple conversions") {
    // Angstrom <-> Bohr
    double a_to_bohr = metatomic_torch::unit_conversion_factor("Angstrom", "Bohr");
    CHECK(a_to_bohr == Approx(1.88972612590485));

    double bohr_to_a = metatomic_torch::unit_conversion_factor("Bohr", "Angstrom");
    CHECK_APPROX_ULP(bohr_to_a, 0.52917721054482);

    // Angstrom <-> nm
    double a_to_nm = metatomic_torch::unit_conversion_factor("Angstrom", "nm");
    CHECK_APPROX_ULP(a_to_nm, 0.1);

    // Angstrom <-> meter
    double a_to_m = metatomic_torch::unit_conversion_factor("Angstrom", "meter");
    CHECK_APPROX_ULP(a_to_m, 1e-10);

    // eV <-> meV
    double ev_to_mev = metatomic_torch::unit_conversion_factor("eV", "meV");
    CHECK_APPROX_ULP(ev_to_mev, 1000.0);

    // eV <-> Hartree
    double ev_to_hartree = metatomic_torch::unit_conversion_factor("eV", "Hartree");
    CHECK(ev_to_hartree == Approx(0.0367493));

    // eV <-> Rydberg
    double ev_to_ry = metatomic_torch::unit_conversion_factor("eV", "Ry");
    CHECK(ev_to_ry == Approx(0.0734987));
}


TEST_CASE("Compound unit expressions") {
    // eV/Angstrom <-> Hartree/Bohr (force)
    double f_conv = metatomic_torch::unit_conversion_factor("eV/Angstrom", "Hartree/Bohr");
    CHECK(f_conv == Approx(0.0194469));

    // kJ/mol <-> kcal/mol (energy)
    double e_conv = metatomic_torch::unit_conversion_factor("kJ/mol", "kcal/mol");
    CHECK_APPROX_ULP(e_conv, 1000.0 / 4184.0);

    // eV/Angstrom^3 (pressure) -- identity
    double p_id = metatomic_torch::unit_conversion_factor("eV/Angstrom^3", "eV/A^3");
    CHECK_APPROX_ULP(p_id, 1.0);
}


TEST_CASE("Fractional power expressions") {
    // (eV*u)^(1/2) should have dimension of momentum: M*L/T
    // Compare to u*A/fs
    double conv = metatomic_torch::unit_conversion_factor("(eV*u)^(1/2)", "u*A/fs");
    // Cross-check: sqrt(eV_SI * u_SI) / (u_SI * A_SI / fs_SI)
    double ev_si = 1.602176634e-19;
    double u_si = 1.66053906660e-27;
    double a_si = 1e-10;
    double fs_si = 1e-15;
    double expected = std::sqrt(ev_si * u_si) / (u_si * a_si / fs_si);
    CHECK(conv == Approx(expected));

    // (eV/u)^(1/2) has dimension of velocity: L/T
    double v_conv = metatomic_torch::unit_conversion_factor("(eV/u)^(1/2)", "A/fs");
    double v_expected = std::sqrt(ev_si / u_si) / (a_si / fs_si);
    CHECK(v_conv == Approx(v_expected));
}


TEST_CASE("Case insensitive unit lookup") {
    double c1 = metatomic_torch::unit_conversion_factor("eV", "hartree");
    double c2 = metatomic_torch::unit_conversion_factor("EV", "HARTREE");
    double c3 = metatomic_torch::unit_conversion_factor("Ev", "Hartree");
    CHECK_APPROX_ULP(c1, c2);
    CHECK_APPROX_ULP(c1, c3);
}


TEST_CASE("Whitespace in unit expressions") {
    double c1 = metatomic_torch::unit_conversion_factor("eV / Angstrom", "Hartree/Bohr");
    double c2 = metatomic_torch::unit_conversion_factor("eV/Angstrom", "Hartree/Bohr");
    CHECK_APPROX_ULP(c1, c2);

    double c3 = metatomic_torch::unit_conversion_factor("( eV * u ) ^ ( 1 / 2 )", "u*A/fs");
    double c4 = metatomic_torch::unit_conversion_factor("(eV*u)^(1/2)", "u*A/fs");
    CHECK_APPROX_ULP(c3, c4);
}


TEST_CASE("Errors") {
    CHECK_THROWS_WITH(
        metatomic_torch::unit_conversion_factor("eV", "Angstrom"),
        Contains(
            "dimension mismatch in unit conversion: 'eV' has dimension "
            "L^2 T^-2 M but 'Angstrom' has dimension L"
        )
    );

    CHECK_THROWS_WITH(
        metatomic_torch::unit_conversion_factor("foobar", "eV"),
        Contains("unknown unit 'foobar'")
    );

    CHECK_THROWS_WITH(
        metatomic_torch::unit_conversion_factor("eV*(", "eV"),
        Contains("unit expression has unbalanced parentheses")
    );

    CHECK_THROWS_WITH(
        metatomic_torch::unit_conversion_factor("ev^eV", "eV"),
        Contains(
            "exponent in unit expression must be dimensionless, "
            "got dimension L^2 T^-2 M for exponent 'eV'"
        )
    );

    SECTION("Overflow") {
        // Test overflow with extreme exponent on unit with small SI factor
        // u (atomic mass unit) has factor 1.66e-27, so u^(-50) overflows
        CHECK_THROWS_WITH(
            metatomic_torch::unit_conversion_factor("u^-50", "u^-50"),
            Contains(
                "unit conversion factor overflows: exponentiation result is "
                "infinite or NaN for 'u ^ -50'"
            )
        );

        // eV has factor 1.6e-19, so eV^(-100) overflows
        CHECK_THROWS_WITH(
            metatomic_torch::unit_conversion_factor("eV^(-100)", "eV^(-100)"),
            Contains(
                "unit conversion factor overflows: exponentiation result is "
                "infinite or NaN for 'eV ^ -100'"
            )
        );

        // Test overflow with multiplication of units with extreme factors
        CHECK_THROWS_WITH(
            metatomic_torch::unit_conversion_factor("u^(-10) * u^(-10)", "u^(-20)"),
            Contains(
                "unit conversion factor overflows: multiplication result is "
                "infinite or NaN for '(u ^ -10) * (u ^ -10)'"
            )
        );

        // Test overflow with division creating extreme factor
        CHECK_THROWS_WITH(
            metatomic_torch::unit_conversion_factor("mol^10 / mol^(-10)", "mol"),
            Contains(
                "unit conversion factor overflows: division result is "
                "infinite or NaN for '(mol ^ 10) / (mol ^ -10)'"
            )
        );
    }
}



TEST_CASE("mol as dimensionless scaling factor") {
    // kJ/mol to eV: kJ_SI / mol_SI / eV_SI
    double conv = metatomic_torch::unit_conversion_factor("kJ/mol", "eV");
    double kj_si = 1000.0;
    double mol_si = 6.02214076e23;
    double ev_si = 1.602176634e-19;
    double expected = (kj_si / mol_si) / ev_si;
    CHECK_APPROX_ULP(conv, expected);
}


TEST_CASE("hbar/Bohr momentum conversion") {
    double conv = metatomic_torch::unit_conversion_factor("hbar/Bohr", "u*A/fs");
    // hbar_SI / Bohr_SI / (u_SI * A_SI / fs_SI)
    double hbar_si = 1.0545718176462e-34;
    double bohr_si = 5.2917721054482e-11;
    double u_si = 1.6605390689252e-27;
    double a_si = 1e-10;
    double fs_si = 1e-15;
    double expected = (hbar_si / bohr_si) / (u_si * a_si / fs_si);
    CHECK_APPROX_ULP(conv, expected);
}


TEST_CASE("Identity conversions") {
    CHECK_APPROX_ULP(metatomic_torch::unit_conversion_factor("eV", "eV"), 1.0);
    CHECK_APPROX_ULP(metatomic_torch::unit_conversion_factor("Angstrom", "Angstrom"), 1.0);
    CHECK_APPROX_ULP(metatomic_torch::unit_conversion_factor("u*A/fs", "u*A/fs"), 1.0);

    CHECK(metatomic_torch::unit_conversion_factor("", "eV") == 1.0);
    CHECK(metatomic_torch::unit_conversion_factor("eV", "") == 1.0);
    CHECK(metatomic_torch::unit_conversion_factor("", "") == 1.0);
}


TEST_CASE("Fractional power accumulated error") {
    // Test that (eV*u)^(1/3) then ^3 equals eV*u within tolerance
    // This verifies the 1e-10 dimension comparison tolerance handles
    // floating-point accumulation from fractional exponents
    CHECK(metatomic_torch::unit_conversion_factor("((eV*u)^(1/3))^3", "eV*u") == Approx(1.0));

    // Test with 1/2 then ^2
    CHECK_APPROX_ULP(metatomic_torch::unit_conversion_factor("((eV*u)^(1/2))^2", "eV*u"), 1.0);

    // Test nested: ((eV^(1/2))^2) should equal eV
    CHECK_APPROX_ULP(metatomic_torch::unit_conversion_factor("(eV^(1/2))^2", "eV"), 1.0);
}


TEST_CASE("Time unit conversions") {
    // second -> femtosecond
    double s_to_fs = metatomic_torch::unit_conversion_factor("s", "fs");
    CHECK_APPROX_ULP(s_to_fs, 1e15);

    // second -> picosecond
    double s_to_ps = metatomic_torch::unit_conversion_factor("second", "ps");
    CHECK_APPROX_ULP(s_to_ps, 1e12);

    // nanosecond -> femtosecond
    double ns_to_fs = metatomic_torch::unit_conversion_factor("ns", "fs");
    CHECK_APPROX_ULP(ns_to_fs, 1e6);

    // microsecond -> nanosecond
    double us_to_ns = metatomic_torch::unit_conversion_factor("us", "ns");
    CHECK_APPROX_ULP(us_to_ns, 1e3);
}


TEST_CASE("Micro sign (U+00B5) handling") {
    double c1 = metatomic_torch::unit_conversion_factor("um", "Angstrom");
    double c2 = metatomic_torch::unit_conversion_factor("µm", "Angstrom");
    CHECK_APPROX_ULP(c1, c2);

    // µs -> ns (microsecond via micro sign)
    double c3 = metatomic_torch::unit_conversion_factor("us", "ns");
    double c4 = metatomic_torch::unit_conversion_factor("µs", "ns");
    CHECK_APPROX_ULP(c3, c4);
}


TEST_CASE("ModelOutput rejects mismatched quantity and unit") {
    // energy quantity with a force unit
    CHECK_THROWS_WITH(
        torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
            "energy", "eV/A", "system", std::vector<std::string>{}, ""
        ),
        Contains(
            "unit 'eV/A' has dimension L T^-2 M which is incompatible "
            "with quantity 'energy' (expected L^2 T^-2 M)"
        )
    );

    // force quantity with an energy unit
    CHECK_THROWS_WITH(
        torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
            "force", "eV", "system", std::vector<std::string>{}, ""
        ),
        Contains(
            "unit 'eV' has dimension L^2 T^-2 M which is incompatible "
            "with quantity 'force' (expected L T^-2 M)"
        )
    );

    // length quantity with a pressure unit
    CHECK_THROWS_WITH(
        torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
            "length", "eV/A^3", "system", std::vector<std::string>{}, ""
        ),
        Contains(
            "unit 'eV/A^3' has dimension L^-1 T^-2 M which is incompatible with "
            "quantity 'length' (expected L)"
        )
    );
}


TEST_CASE("ModelOutput accepts matching quantity and unit") {
    // These should not throw
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "energy", "eV", "system", std::vector<std::string>{}, ""
    );
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "force", "eV/A", "system", std::vector<std::string>{}, ""
    );
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "pressure", "eV/A^3", "system", std::vector<std::string>{}, ""
    );
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "length", "Angstrom", "system", std::vector<std::string>{}, ""
    );
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "momentum", "u*A/fs", "system", std::vector<std::string>{}, ""
    );
    torch::make_intrusive<metatomic_torch::ModelOutputHolder>(
        "velocity", "A/fs", "system", std::vector<std::string>{}, ""
    );
}


// Call the deprecated 3-arg overload without triggering -Wdeprecated-declarations
#if defined(__GNUC__) || defined(__clang__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#endif
static double unit_conversion_factor_previous(const std::string& q, const std::string& f, const std::string& t) {
    return metatomic_torch::unit_conversion_factor(q, f, t);
}
#if defined(__GNUC__) || defined(__clang__)
#pragma GCC diagnostic pop
#endif

TEST_CASE("3-arg API backward compatibility") {
    DeprecationWarningHandler handler;
    torch::WarningUtils::WarningHandlerGuard guard(&handler);
    torch::WarningUtils::set_warnAlways(true);

    double conv = unit_conversion_factor_previous("energy", "eV", "meV");
    CHECK_APPROX_ULP(conv, 1000.0);
}
